### PR TITLE
feat(topology): Add topology workload node sidebar

### PR DESCRIPTION
**Pull request by divyanshiGupta**
**Story:**  https://github.com/janus-idp/backstage-plugins/issues/122

This PR adds a topology workload node sidebar. The sidebar consists of two tabs Details and Resources.
Details tab shows the details of the workloads such as pods status, name, namespace etc. Resources tab shows the resources connected to it such as pods, services , ingresses. Currently the details panel shows specific details for deployments only as other workloads are not yet supported. Once other workloads are added the sidebar will work for these as well but will show common details only and for showing specific details sidebar will need to be updated.

Light Theme

https://user-images.githubusercontent.com/20724543/230070180-992aa7c8-4012-4fb0-a81e-1d5f1875c455.mov

Dark Theme

https://user-images.githubusercontent.com/20724543/230070215-0c9b1252-87c3-4e8a-b653-777737e63b75.mov

### CODE REVIEW

**Comment by debsmita1** [comment]
can we show an empty state here ?
Screenshot 2023-04-04 at 10 29 32 PM

The Creation timestamp on the sidepanel is not clearly visible in the dark theme
Screenshot 2023-04-04 at 10 33 22 PM

**Comment by divyanshiGupta** [comment]
> can we show an empty state here ? Screenshot 2023-04-04 at 10 29 32 PM

Done
Screenshot 2023-04-05 at 4 18 08 PM

**Comment by divyanshiGupta** [comment]
> The Creation timestamp on the sidepanel is not clearly visible in the dark theme Screenshot 2023-04-04 at 10 33 22 PM

Done
Screenshot 2023-04-05 at 4 19 32 PM

File: plugins/topology/src/__fixtures__/workloadNodeData.ts [hunk 1]
```diff
@@ -0,0 +1,3318 @@
+export const workloadNodeData = {
+  id: 'bfb6932f-64c6-4fe3-b283-16ecec8628c0',
+  type: 'workload',
+  label: 'hello-minikube2',
+  resource: {
```
Comment (invincibleJai) [inline plugins/topology/src/__fixtures__/workloadNodeData.ts:5]: will recommend having only part of the mock data that's used in specs like width, height, and data

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyDetailsTabPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+} from '@patternfly/react-core';
+import PodSet from '../../Pods/PodSet';
+import TopologySideBarDetailsItem from './TopologySideBarDetailsItem';
+import { DeploymentModel } from '../../../models';
+import TopologyDeploymentDetails from './TopologyDeploymentDetails';
+import { V1Deployment } from '@kubernetes/client-node';
+import { BaseNode } from '@patternfly/react-topology';
+
+import './TopologyDetailsTabPanel.css';
+import TopologyResourceLabels from './TopologyResourceLabels';
```
Comment (invincibleJai) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyDetailsTabPanel.tsx:16]: nit: keep styles import at end

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyResourcesTabPanel.test.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+import {
+  workloadNode,
+  workloadNode2,
+} from '../../../__fixtures__/workloadNodeData';
+import { BaseNode } from '@patternfly/react-topology';
+import TopologyResourcesTabPanel from './TopologyResourcesTabPanel';
+
+describe('TopologyDetailsTabPanel', () => {
```
Comment (invincibleJai) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyResourcesTabPanel.test.tsx:10]:
```suggestion
describe('TopologyResourcesTabPanel', () => {
```

File: plugins/topology/src/models.ts [hunk 4]
```diff
@@ -74,3 +74,59 @@ export const resourceModels: { [key: string]: GroupVersionKind } = {
   [ModelsPlural.jobs]: JobGVK,
   [ModelsPlural.statefulsets]: StatefulSetGVK,
 };
+
+export const DeploymentModel = {
```
Comment (invincibleJai) [inline plugins/topology/src/models.ts:78]: the purpose for bringing in `...GVK` was to go away from models as used in the console based on the suggestion from Sam. Do we need all of these or can we compute some on the fly?

**Comment by debsmita1** [comment]
Resource badge for the other workloads are missing

Screenshot 2023-04-10 at 11 09 16 PM

**Comment by divyanshiGupta** [comment]
> Resource badge for the other workloads are missing
> 
> Screenshot 2023-04-10 at 11 09 16 PM

@debsmita1 at the time when this PR was raised we didnt have the support for other workloads in topology plugin and it was decided that any specific workload details will be added in separate stories for updating the sidebar wrt to the workload. Anyways if needed I can add the badges for other workloads in this PR.

File: plugins/topology/src/models.ts (hunk 4 above)
Comment (invincibleJai) [reply plugins/topology/src/models.ts:78]: yeah if needed good to keep only those that are needed and if we need other properties then we don't need to have `..GVK` and `...models` both can extend the existing ones themselves.

@spadgett Let us know if you see any issues with this.

**Comment by invincibleJai** [comment]
The Sidebar Cross icon should be scrolled or be fixed with a sticky header but here just the icon stays while the header scrolls up

image

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyResourceLabels.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+import './TopologyResourceLabels.css';
+
+const TopologyResourceLabels: React.FC<{
+  labels: { [key: string]: string };
+  dataTest?: string;
+}> = ({ labels, dataTest }) => {
+  return (
+    <ul className="topology-resource-labels-list" data-testid={dataTest}>
+      {(Object.keys(labels ?? {}) ?? []).map((key: string, index) => (
```
Comment (invincibleJai) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyResourceLabels.tsx:12]:
```suggestion
      {(Object.keys(labels ?? {})).map((key: string, index) => (
```

File: plugins/topology/src/components/Topology/TopologySideBar/TopologyResourcesTabPanel.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+          nodeData.services.map((service: V1Service) => (
+            <li
+              className="item"
+              style={{ flexDirection: 'column' }}
+              key={service.metadata?.uid}
+            >
+              <span>{service.metadata?.name}</span>
+              <ul>
+                {(service.spec?.ports ?? []).map(
+                  ({ name, port, protocol, targetPort }: any) => (
```
Comment (invincibleJai) [inline plugins/topology/src/components/Topology/TopologySideBar/TopologyResourcesTabPanel.tsx:39]:
```suggestion
                  ({ name, port, protocol, targetPort }: V1ServicePort) => (
```

**Comment by divyanshiGupta** [comment]
> 3. Sidebar pod rings don't update (not in sync with topology view)
> 
> image

Fixed.
cc: @invincibleJai 

https://user-images.githubusercontent.com/20724543/232468255-d044f48d-5784-4ba0-98f1-4e49b32d5455.mov

**Comment by invincibleJai** [comment]
@tumido the merge is blocked with the security hotspot by sonarCloud and it is for diff plugin unrelated to this PR https://sonarcloud.io/project/security_hotspots?id=janus-idp_backstage-plugins&pullRequest=226&resolved=false&types=SECURITY_HOTSPOT

Let us know what shall we do.

**Pull request merged**